File: min_stack/min_stack.rb
class MinStack
  def initialize
    @stack = []
    @min_stack = []
  end

  def push(val)
    @stack.push(val)
    @min_stack.push([val, @min_stack.empty? ? val : [val, @min_stack.last[1]].min])
  end

  def pop
    @stack.pop
    @min_stack.pop
  end

  def top
    @stack.last
  end

  def get_min
    @min_stack.last[1]
  end
end

require 'rspec'

RSpec.describe MinStack do
  it 'performs push, pop, top, and get_min correctly' do
    stack = MinStack.new
    stack.push(-2)
    stack.push(0)
    stack.push(-3)
    expect(stack.get_min).to eq(-3)  # Minimum is -3
    stack.pop
    expect(stack.top).to eq(0)       # Top is now 0
    expect(stack.get_min).to eq(-2)  # Minimum is now -2
  end
end

File: min_stack/min_stack.exs
defmodule MinStack do
  defstruct stack: [], min_stack: []

  def new(), do: %MinStack{}

  def push(%MinStack{stack: stack, min_stack: min_stack} = min_stack_struct, val) do
    new_min =
      case min_stack do
        [] -> val
        [min | _] -> min(val, min)
      end

    %MinStack{
      stack: [val | stack],
      min_stack: [new_min | min_stack]
    }
  end

  def pop(%MinStack{stack: [_ | rest_stack], min_stack: [_ | rest_min_stack]}) do
    %MinStack{stack: rest_stack, min_stack: rest_min_stack}
  end

  def top(%MinStack{stack: [top | _]}), do: top

  def get_min(%MinStack{min_stack: [min | _]}), do: min
end
